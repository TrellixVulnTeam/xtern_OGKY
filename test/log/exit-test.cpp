// TODO: fix logging
// XFAIL: *
// RUN: %llvmgcc %s -c -o %t1.ll -S
// RUN: %projbindir/tern-instr < %t1.ll -o %t2
// RUN: llc -o %t2.s %t2-record.bc
// RUN: %gxx -o %t2 %t2.s %ternruntime -lpthread -lrt
// RUN: env TERN_OPTIONS=log_type=bin:output_dir=%t2.outdir ./%t2
// RUN: %projbindir/logprint -bc %t2-analysis.bc %t2.outdir/tid-0.bin -r -v | FileCheck %s.out

#include <sys/time.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <assert.h>
#include <unistd.h>
#include <errno.h>

#define N (2)

void bar() {
  pthread_exit(NULL);
}

void* thread_func(void *arg) {
  bar();
}

int main(int argc, char *argv[], char *env[]) {
  int ret;
  pthread_t th[N];

  for(unsigned i=0; i<N; ++i) {
    ret  = pthread_create(&th[i], NULL, thread_func, (void*)i);
    assert(!ret && "pthread_create() failed!");
  }

  for(unsigned i=0; i<N; ++i)
    pthread_join(th[i], NULL);

  exit(0);
  return 0;
}
