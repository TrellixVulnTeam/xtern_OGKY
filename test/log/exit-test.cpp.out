CHECK: P ins=58  br i1 %22, label %bb6, label %bb8, !dbg !75, !ins_id !90
CHECK: L idx=38 ins=59 call exit(0) noreturn call void @exit(i32 0) noreturn nounwind, !dbg !91, !ins_id !92
CHECK: L idx=39 ins=n/a sync tern_thread_end() turn = 9




